File: built-in plugins dsp/mb-restore-decrackle.cpp
/**
 * MB De-Crackle
 * Category : effect
 * Type     : gate
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Remove vinyl crackle and surface noise from recordings
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_RESTORE_DECRACKLE_H
#define MB_RESTORE_DECRACKLE_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbRestoreDecrackle : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-restore-decrackle";
    static constexpr const char* PLUGIN_NAME    = "MB De-Crackle";
    static constexpr const char* PLUGIN_TYPE    = "gate";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float threshold = 30f;  // range [0, 100]
    float reduction = 0.7f;  // range [0, 1]
    float quality = 2f;  // range [0, 3]
    };

    MbRestoreDecrackle() = default;
    ~MbRestoreDecrackle() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.threshold = std::clamp(params.threshold, 0f, 100f);
        params.reduction = std::clamp(params.reduction, 0f, 1f);
        params.quality = std::clamp(params.quality, 0f, 3f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB De-Crackle
        return input;
    }
};

#endif // MB_RESTORE_DECRACKLE_H

File: built-in plugins dsp/mb-restore-decrackle.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbRestoreDecracklePlugin: PluginDefinition = { id: 'mb-restore-decrackle', slug: 'mb-restore-decrackle', name: 'MB De-Crackle', category: 'effect', type: 'gate' as any, version: '1.0.0', description: 'Remove vinyl crackle and surface noise from recordings', author: 'Max Booster', grade: 'A', parameters: [{ id: 'threshold', name: 'Threshold', type: 'float', defaultValue: 30, minValue: 0, maxValue: 100, automatable: true }, { id: 'reduction', name: 'Reduction', type: 'float', defaultValue: 0.7, minValue: 0, maxValue: 1, automatable: true }, { id: 'quality', name: 'Quality', type: 'float', defaultValue: 2, minValue: 0, maxValue: 3, automatable: false }], defaultPreset: { threshold: 30, reduction: 0.7, quality: 2 } };

export default MbRestoreDecracklePlugin;
